cpp/include/raft/spatial/knn/detail/ann_ivf_flat.cuh

```diff
@@ -0,0 +1,709 @@
+                  cluster_centers,
+                  n_lists_,
+                  dim_,
+                  dataset,
+                  n_rows_,
+                  labels,
+                  metric_type_,
+                  true,
+                  mc_trainset_ccenters_tmp.data(),
+                  mc_trainset_csizes_tmp.data(),
+                  false,
+                  stream_,
+                  &kmeans_mem_res);
+}
+
+template <typename T>
+__global__ void build_index_kernel(const uint32_t* labels,
+                                   const uint32_t* list_offsets,
+                                   const T* dataset,
+                                   T* list_data,
+                                   uint32_t* list_index,
+                                   uint32_t* list_lengths,
+                                   uint32_t n_rows,
+                                   uint32_t dim,
+                                   uint32_t veclen)
+{
+  const int i = blockDim.x * blockIdx.x + threadIdx.x;
+  if (i >= n_rows) { return; }
+
+  auto list_id     = labels[i];
+  auto inlist_id   = atomicAdd(list_lengths + list_id, 1);
+  auto list_offset = list_offsets[list_id];
+
+  // Record the source vector id in the index
+  list_index[list_offset + inlist_id] = i;
+
+  // The data is written in interleaved groups of `WarpSize` vectors
+  using interleaved_group = Pow2<WarpSize>;
+  auto group_offset       = interleaved_group::roundDown(inlist_id);
+  auto ingroup_id         = interleaved_group::mod(inlist_id) * veclen;
+
+  // Point to the location of the interleaved group of vectors
+  list_data += (list_offset + group_offset) * dim;
+
+  // Point to the source vector
+  dataset += i * dim;
+
+  // Interleave dimensions of the source vector while recording it.
+  for (uint32_t l = 0; l < dim; l += veclen) {
+    for (uint32_t j = 0; j < veclen; j++) {
+      list_data[l * WarpSize + ingroup_id + j] = dataset[l + j];
+    }
+  }
+}
+
+template <typename T>
+void cuivflHandle<T>::cuivflBuildIndex(const T* dataset,
+                                       T* trainset,
+                                       uint32_t n_rows,
+                                       uint32_t n_rows_train)
+{
+  n_rows_ = n_rows;
+  RAFT_EXPECTS(n_rows_ > 0, "empty dataset");
+
+  rmm::mr::managed_memory_resource managed_memory;
+  rmm::device_uvector<float> centriod_managed_buf(n_lists_ * dim_, stream_, &managed_memory);
+  auto centriod_managed_ptr = centriod_managed_buf.data();
+
+  static_assert(std::is_same_v<T, float> || std::is_same_v<T, uint8_t> || std::is_same_v<T, int8_t>,
+                "unsupported data type");
+
+  // Alloc manage memory for centriods, trainset and workspace
+  rmm::device_uvector<uint32_t> labels_buf(n_rows_, stream_, &managed_memory);  // [numDataset]
+  auto labels = labels_buf.data();
+
+  // Predict labels of the whole dataset
+  cuivflBuildOptimizedKmeans(centriod_managed_ptr, dataset, trainset, labels, n_rows, n_rows_train);
```

Comment:
Same here- why not just train the clusters on the train set?

Reply:
Hm, we use the `dataset` in the last step of this function, which is to label the dataset and update the cluster centers to be in the centers of whole `dataset` (rather than the smaller `trainset`). In addition, we use the labels later for building the index.

Thanks for the good points, I'll add these answers to the documentation.